staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -379,10 +375,10 @@ func WaitFor(wait WaitFunc, fn ConditionFunc, done <-chan struct{}) error {
 				return nil
 			}
 			if !open {
-				return ErrWaitTimeout
+				break FOR
 			}
 		case <-done:
-			closeCh()
+			break FOR
```

Comment:
I think its easier to read if we just return here. Also can we add a test for this case?

Reply:
TestWaitForWithClosedChannel and TestWaitForWithEarlyClosingWaitFunc test two branches which return ErrWaitTimeout.


Also I document the case when `down` and `c` are closed:
```
// Be careful, when the 'wait' func returns a closed channel and the 'done' is also closed,
// the behavior of this function depends on the golang 'select'.
```

Reply:
For the test, I think cheftako wants to make sure that the close of the `done` channel can stop the `WaitFor`, regardless of the `WaitFunc`. For example, if the `WaitFunc` doesn't send anything to the its returned channel, nor close the channel, `WaitFor` should still be stopped by the close of the `done` channel. Can you add such a case?

Reply:
Done.

Reply:
I would like to see an additional test something like 

        stopCh := make(chan struct{})
        go func() {
             time.Sleep(time.Second)
             close(stopCh)
        }
	start := time.Now()
	err := WaitFor(poller(ForeverTestTimeout, ForeverTestTimeout), func() (bool, error) {
		return false, nil
	}, stopCh)

This clearly tests only the stopCh channel being closed and does not rely on it starting closed.